<a href="https://colab.research.google.com/github/Slalani555/Deep-Learning-Python/blob/main/Deep_Learning_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **An example of neural network that uses the Python Library Keras to learn to classify handwritten digits.**

- Problem Statement- Classify grayscale images of handwritten digits(28x28 pixels)into their categories (0-9).

- We are using MNIST dataset which consists of 60,000 training images and 10,000 test images.

- This dataset has been assembled by the National Institute of Standards and Technology in the 1980s.

In [2]:
from tensorflow.keras.datasets import mnist
(train_images,train_labels),(test_images,test_labels)=mnist.load_data()

11501568/11490434 [==============================] - 0s 0us/step


*The images are encoded as NumPy arrays and the labels are an array of digits, ranging from 0 to 9. There is a one-to-one corresposndence between the images and labels.*

In [3]:
train_images.shape

(60000, 28, 28)

In [4]:
len(train_labels)

60000

In [5]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [6]:
test_images.shape

(10000, 28, 28)

In [7]:
len(test_labels)

10000

In [8]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

*The network architecture*

In [9]:
from tensorflow import keras
from tensorflow.keras import layers
model=keras.Sequential([layers.Dense(512,activation='relu'),layers.Dense(10,activation='softmax')])

*The compilation step*

In [10]:
model.compile(optimizer="rmsprop",
              loss="sparse_categorical_crossentropy",
              metrics=["accuracy"])

*Preprocessing-Preparing the image data*

- Reshape it into the shape the model expects and scaling it so that all the values are in the interval [0,1]
- Previously, our training images were stored in an array of shape (60000,28,28) of type uint8 with values in the [0,255] interval. Let's transform it into a float32 array of shape (60000,28*28) with values between 0 and 1.

In [11]:
train_images=train_images.reshape((60000,28*28))
train_images=train_images.astype("float32")/255
test_images=test_images.reshape((10000,28*28))
test_images=test_images.astype("float32")/255

*Fitting the model*

In [12]:
model.fit(train_images,train_labels, epochs=5,batch_size=128)

Epoch 1/5
469/469 [==============================] - 5s 9ms/step - loss: 0.2578 - accuracy: 0.9247
Epoch 2/5
469/469 [==============================] - 4s 9ms/step - loss: 0.1027 - accuracy: 0.9690
Epoch 3/5
469/469 [==============================] - 4s 9ms/step - loss: 0.0680 - accuracy: 0.9800
Epoch 4/5
469/469 [==============================] - 4s 9ms/step - loss: 0.0501 - accuracy: 0.9848
Epoch 5/5
469/469 [==============================] - 4s 9ms/step - loss: 0.0375 - accuracy: 0.9890


*Using the model to make predictions*

In [13]:
test_digits=test_images[0:10]
predictions=model.predict(test_digits)
predictions[0]

array([3.9508175e-09, 1.4951948e-12, 3.8224732e-07, 8.2494080e-06,
       1.5632235e-14, 5.1348628e-09, 2.4184699e-16, 9.9999130e-01,
       1.4333632e-09, 1.1071283e-07], dtype=float32)

Observation: Each number of index i in that array corresponds to the probability that digit image test_digits[0] belongs to class i. This first test digit has the high probability core at index 7, so according to our model it must be a 7.

In [14]:
predictions[0].argmax()

7

In [15]:
predictions[0][7]

0.9999913

In [16]:
test_labels[0]

7

The above line of code checks whether the test label agrees with our prediction.

In [25]:
predictions[1]

array([1.6625302e-11, 3.1176370e-07, 9.9999964e-01, 3.8809276e-09,
       3.5756963e-19, 8.6187821e-11, 8.0548190e-12, 9.0319256e-18,
       4.4833062e-10, 3.7441149e-17], dtype=float32)

In [26]:
predictions[1].argmax()

2

In [27]:
predictions[1][2]

0.99999964

In [28]:
test_labels[1]

2

*Evaluating our model on unseen data*

In [29]:
test_loss,test_accuracy=model.evaluate(test_images,test_labels)
print(f"test_accuracy:  {test_accuracy}")

313/313 [==============================] - 1s 3ms/step - loss: 0.0749 - accuracy: 0.9778
test_accuracy:  0.9778000116348267


Observations:- The test set accuracy turns out to be 97.78% which is a bit lower than out training set accuracy (98.9%).